registration/round.go

```diff
@@ -0,0 +1,265 @@
+func (r *round) submit(ctx context.Context, key, challenge []byte) (<-chan error, error) {
+	r.batchMutex.Lock()
+	defer r.batchMutex.Unlock()
+
+	if r.members+len(r.pendingSubmits) >= r.maxMembers {
+		return nil, ErrMaxMembersReached
+	}
+
+	if r.batch == nil {
+		r.pendingSubmits = make(map[string]pendingSubmit)
+		r.batch = leveldb.MakeBatch(r.maxBatchSize)
+		time.AfterFunc(r.flushInterval, r.flushPendingSubmits)
+	}
+
+	pending, ok := r.pendingSubmits[string(key)]
+	if ok {
+		switch {
+		case bytes.Equal(challenge, pending.challenge):
+			return nil, fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
+		default:
+			return nil, ErrConflictingRegistration
+		}
+	}
+
+	registered, err := r.db.Get(key, nil)
+	switch {
+	case errors.Is(err, leveldb.ErrNotFound):
+
+		// OK - challenge is not registered yet.
+	case err != nil:
+		return nil, fmt.Errorf("failed to check if challenge is registered: %w", err)
+	case bytes.Equal(challenge, registered):
+		return nil, fmt.Errorf("%w: key: %X", ErrChallengeAlreadySubmitted, key)
+	default:
+		return nil, ErrConflictingRegistration
+	}
+
+	r.batch.Put(key, challenge)
+	done := make(chan error, 1)
+	r.pendingSubmits[string(key)] = pendingSubmit{
+		done,
+		challenge,
+	}
+
+	if r.batch.Len() >= r.maxBatchSize {
+		r.flushPendingSubmitsLocked()
+	}
+
+	return done, nil
+}
+
+func (r *round) flushPendingSubmits() {
+	r.batchMutex.Lock()
+	defer r.batchMutex.Unlock()
+	r.flushPendingSubmitsLocked()
+}
+
+func (r *round) flushPendingSubmitsLocked() {
+	if r.batch == nil {
+		return
+	}
+	logging.FromContext(context.Background()).
+		Debug("flushing pending submits", zap.Int("num", len(r.pendingSubmits)), zap.Uint("round", r.epoch))
+	start := time.Now()
+	err := r.db.Write(r.batch, &opt.WriteOptions{Sync: true})
+	if err == nil {
+		batchWriteLatencyMetric.Observe(time.Since(start).Seconds())
+	}
+	for _, pending := range r.pendingSubmits {
+		pending.done <- err
+		close(pending.done)
+	}
+	r.members += len(r.pendingSubmits)
+	r.membersCounter.Add(float64(len(r.pendingSubmits)))
+
+	r.pendingSubmits = nil
+	r.batch = nil
+}
+
+func countMembersInDB(db *leveldb.DB) (count int) {
+	iter := db.NewIterator(nil, nil)
+	defer iter.Release()
+	for iter.Next() {
+		count++
+	}
+	return count
+}
+
+func (r *round) calcMembershipRoot() ([]byte, error) {
+	r.flushPendingSubmits()
+	mtree, err := merkle.NewTreeBuilder().
+		WithHashFunc(shared.HashMembershipTreeNode).
+		Build()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize merkle tree: %v", err)
+	}
+
+	iter := r.db.NewIterator(nil, nil)
+	defer iter.Release()
+	for iter.Next() {
+		if err := mtree.AddLeaf(iter.Value()); err != nil {
+			return nil, err
+		}
+	}
+
+	return mtree.Root(), nil
+}
+
+func (r *round) getMembers() (members [][]byte) {
+	iter := r.db.NewIterator(nil, nil)
+	defer iter.Release()
+	for iter.Next() {
+		member := make([]byte, len(iter.Value()))
+		copy(member, iter.Value())
+		members = append(members, member)
```

Comment:
NIT append also copies if used like this and avoids the double call to `iter.Value()`:

```suggestion
		var member []byte
		member = append(member, iter.Value()...)
		members = append(members, member)
```

This is results in a copy as can be seen here: https://go.dev/play/p/orMHwAnr48w